keps/sig-apps/0003-kubernetes-application-api.md

```diff
@@ -0,0 +1,1120 @@
+              key: password
+        ports:
+        - containerPort: 80
+          name: wordpress
+        volumeMounts:
+        - name: wordpress-persistent-storage
+          mountPath: /var/www/html
+  volumeClaimTemplates:
+       - metadata:
+           name: mysql-persistent-storage
+         spec:
+           accessModes: [ "ReadWriteOnce" ]
+           resources:
+             requests:
+               storage: 250Gi
+```
+
+Above we generate two Applications, one for the MySQL RDBMs and another for the WordPress web server. Each of these 
+Applications can be managed independently. For instance, if the WordPress web server is deleted, the components of the 
+MySQL RDBMS will not be affected. However, the dependency between the two workloads is made explicit in the 
+`Spec.Dependencies` of the `wordpress` Application. Tools and UIs that examine the `wordpress` application are made 
+aware of its dependency on the `mysql` Application, and they can surface the dependency to users.
+
+#### Selecting Components
+In order to find the `Spec.Componenets` of an Application, a client should perform the following.
+
+1. For each GroupVersionKind in the Applications `Spec.Components`, list all objects of that Kind, in the namespace of 
+the Application, that match the Application's `Spec.Selector`.
+1. For each selected object, validate that the object has a `kubernetes.io/application` that matches the name of the 
+Application. Discard any objects that do not have such an annotation. This ensures that selector overlap does not 
+violate the intentions of the application creators and administrators.
+1. Unless otherwise indicated, clients should ensure that the `OwnerReferences` of the object contain the `UID` of the 
+Application.
+
+#### Retrieving Applications
+In order to retrieve the Application corresponding to an API object, a client should perform the following.
+
+1. Examine the object for a `kubernetes.io/application` annotation.
+1. If no annotation is found, the object is not associated with an Application.
+1. If the annotation is found, retrieve the Application object with the indicated name from the same namespace of the 
+object.
+1. If no such Application is found, the Application is either deleted or has not yet been created.
+1. If the Application is found, validate that the `Spec.Selector` of the Application object matches the labels of the 
+object. If the object's labels do not match, the object has been orphaned.
+1. Check the `OwnerReferences` of the object for a reference to the Application by `UID` to ensure that ownership has 
+been established.
+
+#### Retrieving Dependencies
+In order to retrieve the `Spec.Dependencies` of an Application, a client should perform the following.
+
+1. For each name in the `Spec.Dependencies` list, retrieve the Application of that name, in the namespace of the 
+Application under consideration. If no such Application is found, the dependency is not installed.
+
+#### Health Checks
```

Comment:
This is my favorite feature of the entire proposal.

Reply:
I really think health checks are othogonal to applications and should be represented either related to a service (so as to provide a concrete attachment to something we already have) or be completely orthogonal (define "health" of a service as a minimum number of healthy endpoints).  I don't want to couple them to something generic, I'd want to have them be a useful tool even if you don't use applications.

Reply:
Is there a summary of user feedback on `helm test` anywhere? 

https://github.com/kubernetes/helm/blob/master/docs/chart_tests.md

Reply:
I updated this to use a Deployment to simplify the controller logic as was suggested. I added an indication that we will add Service based health checks when detecting the health of a Service based on its available endpoints becomes a core feature.

Reply:
I'm mostly agnostic about implementation details, but I am a big `+1` on the basic idea of having a kube-native concept for "do not actually try to run this pod unless this other set of resources I provide is passing health checks"